src/interface/variable_pack.hpp

```diff
@@ -34,33 +40,96 @@ class CellVariable;
 
 // some convenience aliases
 namespace vpack_types {
-template <typename T>
-using VarList = std::forward_list<std::shared_ptr<CellVariable<T>>>;
+
 // Sparse and/or scalar variables are multiple indices in the outer view of a pack
 // the pairs represent interval (inclusive) of those indices
 using IndexPair = std::pair<int, int>;
+
 // Flux packs require a set of names for the variables and a set of names for the strings
 // and order matters. So StringPair forms the keys for the FluxPack cache.
 using StringPair = std::pair<std::vector<std::string>, std::vector<std::string>>;
+
 } // namespace vpack_types
 
-// using PackIndexMap = std::map<std::string, vpack_types::IndexPair>;
+// helper class to make lists of variables with labels
+template <typename T>
+class VarListWithLabels {
+ public:
+  VarListWithLabels<T>() = default;
+
+  // Adds a variable to the list if one of the following is true:
+  // a) The variable is not sparse
+  // b) The set of sparse_ids is empty
+  // c) The sparse id of the variable is contained in the set of sparse_ids
+  void Add(const std::shared_ptr<CellVariable<T>> &var,
+           const std::unordered_set<int> &sparse_ids = {}) {
+    if (!var->IsSparse() || sparse_ids.empty() ||
+        (sparse_ids.count(var->GetSparseID()) > 0)) {
+      vars_.push_back(var);
+      // The label here is used in the key for variable pack caching, so it has to include
+      // the allocation status of the variable, since we cannot reuse a variable pack if
+      // one of its variables has a changed allocation status
+      labels_.push_back(var->label() + (var->IsAllocated() ? "_A" : ""));
+    }
+  }
+
+  const auto &vars() const { return vars_; }
+  const auto &labels() const { return labels_; }
+
+ private:
+  CellVariableVector<T> vars_;
+  std::vector<std::string> labels_;
+};
+
+// using PackIndexMap = std::unordered_map<std::string, vpack_types::IndexPair>;
 class PackIndexMap {
  public:
   PackIndexMap() = default;
-  vpack_types::IndexPair &operator[](const std::string &key) {
-    if (!map_.count(key)) {
-      map_[key] = vpack_types::IndexPair(0, -1);
+
+  auto &get(const std::string &key) {
+    auto itr = map_.find(key);
+    if (itr == map_.end()) {
+      auto err = "PackIndexMap does not have key '" + key + "'";
+      PARTHENON_THROW(err.c_str());
```

Comment:
I've changed this to:
```c++
  // This is dangerous! Use at your own peril!
  // It will silently return invalid indices if the key doesn't exist (e.g. misspelled or
  // sparse id not part of label)
  const vpack_types::IndexPair &operator[](const std::string &key) const {
    static const vpack_types::IndexPair invalid_indices(-1, -1);
    auto itr = map_.find(key);
    if (itr == map_.end()) {
      return invalid_indices;
    }

    return itr->second;
  }
```